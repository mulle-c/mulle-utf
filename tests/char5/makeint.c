#include "dependencies.h"

#include <stdio.h>
#include <stdlib.h>


static void   bit32( char *s)
{
   uint32_t   encoded;
   char       decoded[ 6];
   size_t     len;
   size_t     decoded_len;

   len = strlen( s);
   if( ! mulle_char5_is_char5string32( s, len))
      return;

   encoded     = mulle_char5_encode32( s, len);
   decoded_len = mulle_char5_decode32( encoded, decoded, sizeof( decoded));

   if( decoded_len != len || strncmp( decoded, s, decoded_len))
   {
      printf( "%s failed\n", s);
      abort();
   }
   printf( "32bit: %.*s (%d): 0x%lx\n", (int) decoded_len, decoded, decoded_len, (unsigned long) encoded);
}


static void   bit64( char *s)
{
   uint64_t   encoded;
   char       decoded[ 12];
   size_t     len;
   size_t     decoded_len;

   len = strlen( s);
   if( ! mulle_char5_is_char5string64( s, len))
      return;

   encoded     = mulle_char5_encode64( s, len);
   decoded_len = mulle_char5_decode64( encoded, decoded, sizeof( decoded));
   if( decoded_len != len || strncmp( decoded, s, decoded_len))
   {
      printf( "%s failed\n", s);
      abort();
   }
   printf( "64bit: %.*s (%d): 0x%llx\n", (int) decoded_len, decoded, decoded_len, (unsigned long long) encoded);
}


static void  test( char *s)
{
   bit32( s);
   bit64( s);
}


int  main()
{
   int  c;
   int  i;
   int  d;

   test( "");
   test( "a");
   test( "ab");
   test( "abc");
   test( "abcd");
   test( "abcde");
   test( "abcdea");
   test( "abcdeab");
   test( "abcdeabc");
   test( "abcdeabcd");
   test( "abcdeabcde");
   test( "abcdeabcdea");
   test( "abcdeabcdeab");
   test( "abcdeabcdeabc");  // 13

   return( 0);
}

