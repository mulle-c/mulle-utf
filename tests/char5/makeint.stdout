32bit:  (0): 0x0
64bit:  (0): 0x0
32bit: A (1): 0x5
64bit: A (1): 0x5
32bit: AA (2): 0xa5
64bit: AA (2): 0xa5
32bit: AAA (3): 0x14a5
64bit: AAA (3): 0x14a5
32bit: AAAA (4): 0x294a5
64bit: AAAA (4): 0x294a5
32bit: AAAAA (5): 0x5294a5
64bit: AAAAA (5): 0x5294a5
32bit: AAAAAA (6): 0xa5294a5
64bit: AAAAAA (6): 0xa5294a5
64bit: AAAAAAA (7): 0x14a5294a5
64bit: AAAAAAAA (8): 0x294a5294a5
64bit: AAAAAAAAA (9): 0x5294a5294a5
64bit: AAAAAAAAAA (10): 0xa5294a5294a5
64bit: AAAAAAAAAAA (11): 0x14a5294a5294a5
64bit: AAAAAAAAAAAA (12): 0x294a5294a5294a5
